### ISSUE TITLE
Basic grid refinement 

### ISSUE THREAD

**Issue by jbcaillau** [control-toolbox/CTDirect.jl#119]
@PierreMartinon now that it is possible to pass a `time_grid`, we can rather easily combine solve with grid refinement and warm start to solve iteratively on a adaptive discretisation.

**Comment by ocots** [comment]
@PierreMartinon @jbcaillau @0yassine0 In the [Space shuttle problem](https://github.com/0Yassine0/COTS.jl/blob/eb8d62e021e5bb90c556afd8a3b6691fcf591f95/Problems/JuMP/space_Shuttle_JMP.jl#L59) from `JuMP` tutorial, the time steps are considered as variables to optimize. Do we agree that we cannot do that for the moment?

**Comment by PierreMartinon** [comment]
The whole time grid (t_i)_i=0..n meaning possibly non-uniform steps, or just a free final time with a regular subdivision ?
We can do a
- a fixed uniform time grid (set grid_size=...) with fixed or free t0,tf.
- a fixed non-uniform time grid, also compatible with fixed or free t0,tf. More precisely the normalized grid in [0,1] is the one fixed, so if t0 and/tf are free, all the time steps can change, but the 'repartition' of the steps is fixed.
We cannot do
- optimize the whole vector [t0, t1, ..., tf] with each t_i an independant variable

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @0Yassine0 yes, the usual manual trick works to glue things (*e.g.* multidynamics with unknown switching time). Actually, it is interesting and not completely standard to optimise the grid points themselves: interesting to compare with a uniform grid size (should be better, but price to pay = more vars).

@ocots I remember that you did sth similar, no?

@PierreMartinon @joseph-gergaud not very difficult to add this possibility in the transcription process (enforcing a lower bound on each $t_{i+1}-t_i$ to avoid stupid things), provided there is a demonstrated advantage.

**Comment by PierreMartinon** [comment]
I think I can do a simple example for this. @jbcaillau, you typically take positive Dt_i instead of the t_i themselves, easier constraint ;-)

**Comment by PierreMartinon** [comment]
> @PierreMartinon now that it is possible to pass a `time_grid`, we can rather easily combine solve with grid refinement and warm start to solve iteratively on a adaptive discretisation.

A basic illustration here https://github.com/control-toolbox/CTDirect.jl/pull/158
(on uniform grids, but still passed as explicit grids...)

**Comment by PierreMartinon** [comment]
> @ocots @PierreMartinon @0Yassine0 yes, the usual manual trick works to glue things (_e.g._ multidynamics with unknown switching time). Actually, it is interesting and not completely standard to optimise the grid points themselves: interesting to compare with a uniform grid size (should be better, but price to pay = more vars).
> 
> @ocots I remember that you did sth similar, no?
> 
> @PierreMartinon @joseph-gergaud not very difficult to add this possibility in the transcription process (enforcing a lower bound on each ti+1−ti to avoid stupid things), provided there is a demonstrated advantage.

Added an example in https://github.com/control-toolbox/CTDirect.jl/pull/158

**Comment by jbcaillau** [comment]
@PierreMartinon as discussed tonight: could be nice to have a specific option that replaces a fixed `grid_size` by a more or less automated grid refinement (list of sizes...)

**Comment by jbcaillau** [comment]
see also https://dl.acm.org/doi/abs/10.1137/0715004 (grid balancing rather than refinement - to be checked, ref from @rveltz)

@ocots please add the ref to Cash *et al*
